### DISCUSSION TITLE
Document editor progress

### DISCUSSION THREAD

**Discussion by VendelinSlezak** [VendelinSlezak/BleskOS#71] [Show and tell]
Here I will show progress in developing program Document editor. Actual version is able to open ODT files in very simple way:

![Snímka obrazovky_2024-02-17_18-32-36](https://github.com/VendelinSlezak/BleskOS/assets/43180618/acf7e7ef-d478-4acf-9b94-c7a0451ee373)

From initial version there are multiple improvements, from which most visible is wrapping of lines and ability to select and copy text to normal text area:

![Snímka obrazovky_2024-02-17_18-33-31](https://github.com/VendelinSlezak/BleskOS/assets/43180618/916e990d-533c-4b04-b247-c82975dcb81e)

**Comment by VendelinSlezak** [comment]
In update 18 there are many improvements. BleskOS can decode different font size, emphasis, color and background color.

![Snímka obrazovky_2024-02-19_21-23-17](https://github.com/VendelinSlezak/BleskOS/assets/43180618/db9a2afc-787f-4c2e-99dd-b775420128db)

**Comment by VendelinSlezak** [comment]
Update 19 almost reached full list of supported features of documents for opening ODT files. Only thing that is not supported yet is different size of pages. And some of implemented features are sometimes not displayed correctly, what will be fixed in next updates.

![Snímka obrazovky_2024-02-20_13-57-13](https://github.com/VendelinSlezak/BleskOS/assets/43180618/b78eaa58-95db-4c6c-aa93-80e26b690c98)

**Reply by defdefred** [reply]
could you share your test.odt?

**Comment by VendelinSlezak** [comment]
Document editor in update 20 can parse everything from ODT files that will be supported for now and there is also basic support for DOCX files.

![Snímka obrazovky_2024-02-22_20-14-02](https://github.com/VendelinSlezak/BleskOS/assets/43180618/0973d70c-9b24-4e63-9ca7-473d5dd705c2)

![Snímka obrazovky_2024-02-22_20-13-52](https://github.com/VendelinSlezak/BleskOS/assets/43180618/c0792e2b-1dc5-4f6b-b3e1-f9e84aa8de1b)

[Test.odt](https://github.com/VendelinSlezak/BleskOS/files/14378499/Test.odt)
[Test.docx](https://github.com/VendelinSlezak/BleskOS/files/14378502/Test.docx)

**Comment by VendelinSlezak** [comment]
Update 21 can open DOCX files on same level as ODT files

![Snímka obrazovky_2024-02-24_15-00-15](https://github.com/VendelinSlezak/BleskOS/assets/43180618/b2ef41ec-7eac-4d40-a19e-c47f7b0038e8)

Next goal is to be able to convert back and save document to ODT or DOCX. There will be also possibility to use BleskOS memory format.

**Comment by VendelinSlezak** [comment]
Update 22 brings creating ZIP files and update 23 starts with work on converting Document Memory Format to ODT file. So far it can convert only paragraphs, but it work and result can be opened in normal office program.

**Comment by VendelinSlezak** [comment]
From Update 29 is Document editor able to successfully save Test file as .odt that can be opened by normal office program. Below is result in LibreOffice Writer:

![Snímka obrazovky_2024-03-14_12-45-26](https://github.com/VendelinSlezak/BleskOS/assets/43180618/deba2472-3fd4-476a-ab7d-12c8dc30960b)
